--> A rule to generate a bed from mutect2 vcf, on tumor versus normal with panel of normals

- input files : 
VCF file containing the variants identified in the tumor sample compared to the normal sample
```
        Mutect2_vcf = "Mutect2_TvNp/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_twicefiltered_TvNp.vcf.gz"
```
- output files :
Variant BED file:  the output BED file that contains the genomic regions of the filtered variants from the Tumor-Normal VCF file. The BED file include the chromosome, start position, end position, and any additional relevant information for each variant.
```
        BED = temp("variant_bed_TvN/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_TvN.bed")
```
-->  Run samtools mpileup, on tumor versus normal with panel of normals

his rule performs variant calling using samtools mpileup on the tumor and normal samples with the help of a panel of normals (PON). samtools mpileup generates a pileup file that represents the alignment and coverage information for each position in the genome.
- input files :
-> Tumor BAM file: The aligned reads in BAM format for the tumor sample.
-> Normal BAM file: The aligned reads in BAM format for the normal sample.
-> Panel of Normals (PON) BAM file: The aligned reads in BAM format for the panel of normal samples.
```
        BED = "variant_bed_TvNp/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_TvNp.bed",
        BAM = "bam/{tsample}.nodup.recal.bam" if config["remove_duplicates"] == True else "bam/{tsample}.recal.bam",
        BAI = "bam/{tsample}.nodup.recal.bam.bai" if config["remove_duplicates"] == True else "bam/{tsample}.recal.bam.bai"
```
- output files :
-> Pileup file: The output file generated by samtools mpileup containing the alignment and coverage information for each position in the genome.
```
        PILEUP = temp("pileup_TvN/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_TvN.pileup.gz")
```
- Reference genome: 
The reference genome in FASTA format used for alignment and variant calling.

-->  A rule to split mutect2 results in pieces

This rule splits the Mutect2 results file into smaller pieces to facilitate downstream analysis or processing of the data. By splitting the file, you can work with smaller subsets of variants at a time, improving computational efficiency or enabling parallelization.

- input files : 
Mutect2 VCF file: The Mutect2 variant call format (VCF) file containing the results of the variant calling analysis.
```
        Mutect2_vcf = "Mutect2_TvNp/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_twicefiltered_TvNp.vcf.gz",
        vcf_index = "Mutect2_TvNp/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_twicefiltered_TvNp.vcf.gz.tbi"
```
- output files :
-> Interval-specific VCF file (with bcftools suffix): The output VCF file specific to the interval, compressed using bcftools or similar tool.
-> Interval-specific VCF file: The output VCF file specific to the interval, without any additional suffix or compression.
```
        interval_vcf_bcftools = temp("Mutect2_TvNp_oncotator_tmp/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_TvNp_ON_{interval}_bcftools.vcf.gz"),
        interval_vcf          = temp("Mutect2_TvNp_oncotator_tmp/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_TvNp_ON_{interval}.vcf.gz")
```

--> A rule to annotate mutect2 tumor versus normal and panel of normal results with oncotator  
rule oncotator_pon :
This rule performs annotation of Mutect2 panel of normals (PON) results using Oncotator. Oncotator is a tool used for comprehensive annotation of genomic variants.
- input files : 
Mutect2 VCF file (panel of normals): The Mutect2 variant call format (VCF) file containing the results of the panel of normals comparison.
```
        interval_vcf = "Mutect2_TvNp_oncotator_tmp/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_TvNp_ON_{interval}.vcf.gz"
```
- output files :
The output MAF file with Mutect2 analysis results annotated by Oncotator.
```
        MAF = temp("oncotator_TvNp_tmp/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_ON_{interval}_annotated_TvNp.TCGAMAF")
```

--> concatenate oncotator TvN_pon
rule concatenate_oncotator_pon:
This rule concatenates multiple Oncotator-annotated Mutect2 panel of normals (PON) VCF files into a single consolidated VCF file.
- input files : 
Oncotator-annotated Mutect2 PON VCF files: Multiple VCF files containing the results of Mutect2 panel of normals comparison, annotated with Oncotator.
```
        maf = expand("oncotator_TvNp_tmp/{{tsample}}_Vs_{{nsample}}_PON_{{panel_of_normal}}_ON_{mutect_interval}_annotated_TvNp.TCGAMAF", mutect_interval=mutect_intervals)
```
- output files :
-> Concatenated Oncotator-annotated PON VCF file: The output VCF file containing the consolidated Mutect2 PON results annotated with Oncotator.
-> Temporary file list: The output file containing a list of temporary files generated during the analysis pipeline.
```
        concatened_oncotator = temp("oncotator_TvNp/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_annotated_TvNp.TCGAMAF"),
        tmp_list = temp("oncotator_TvNp_tmp/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_TvNp_oncotator_tmp.list")
```

--> A rule to simplify oncotator output on tumor vs normal samples with panel of normal
rule oncotator_reformat_TvN_pon:
Filter out variants present in both the tumor and the panel of normal samples.
- input files : 
Tumor sample variant calls obtained from Oncotator / Oncotator output in Mutation Annotation Format (MAF) format
```
        maf="oncotator_TvNp/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_annotated_TvNp.TCGAMAF"
```
- output files :
Simplified variant calls specific to the tumor sample.
```
        maf = "oncotator_TvNp_maf/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_TvNp_selection.TCGAMAF",
        tsv = temp("oncotator_TvNp_tsv/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_TvNp.tsv")
```

--> A rule to simplify oncotator output on tumor vs normal samples with panel of normal
rule oncotator_with_pileup_TvN_pon:
The rule "oncotator_with_pileup_TvN_pon" implies a specific approach or procedure involving the use of Oncotator and pileup data for tumor vs. normal (TvN) comparison with a panel of normals (PoN). 
- input files :
-> TSV file contains relevant data or annotations related to the tumor sample, normal sample, and the panel of normal samples used for comparison
-> The pileup file contains information about read coverage and genetic variants at each genomic position for the tumor sample and the normal sample.
```
        tsv = "oncotator_TvNp_tsv/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_TvNp.tsv",
        pileup = "pileup_TvN/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_TvN.pileup.gz"
```
- output files :
Output purpose is to store the combined information from Oncotator annotations and the pileup data, specifically for the analysis of tumor vs. normal (TvN) comparison with a panel of normal (PoN).
```
        tsv = temp("oncotator_TvNp_tsv_pileup/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_TvNp_with_pileup.tsv")
```

--> A rule to simplify oncotator output on tumor vs normal samples with panel of normal
rule oncotator_with_COSMIC_TvN_pon:
the rule aims to simplify the Oncotator output for the analysis of tumor vs. normal samples with a panel of normal, with a particular focus on integrating data from the COSMIC (Catalogue of Somatic Mutations in Cancer) database. This rule streamlines the output by consolidating relevant information and annotations specific to the tumor vs. normal comparison.
- input files : 
```
     tsv = "oncotator_TvNp_tsv_pileup/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_TvNp_with_pileup.tsv"
```
- output files :
```
        tsv = "oncotator_TvNp_tsv_COSMIC/{tsample}_Vs_{nsample}_PON_{panel_of_normal}_TvNp_with_COSMIC.tsv"
```

